File: src/tts/speakableItem.ts
const SPEECH_LANGUAGE_BY_CONTENT_LANGUAGE: Record<string, string> = {
  en: 'en-US',
  he: 'he-IL',
};

const HEBREW_CHARACTER_PATTERN = /[\u0590-\u05FF]/;

export type SpeakableItem = {
  text: string;
  authorLabel: string;
  lang: string;
};

export const getSpeakableItemLanguage = (
  item: Pick<SpeakableItem, 'text' | 'lang'>,
): string | undefined => {
  if (HEBREW_CHARACTER_PATTERN.test(item.text)) {
    return 'he-IL';
  }

  return SPEECH_LANGUAGE_BY_CONTENT_LANGUAGE[item.lang.toLowerCase()];
};

export const getSpeakableItemText = (item: Pick<SpeakableItem, 'authorLabel' | 'text'>): string => {
  const authorPrefix = item.authorLabel ? `At ${item.authorLabel}. ` : '';

  return `${authorPrefix}${item.text}`.trim();
};

File: android/app/src/main/kotlin/com/yonisirote/readmyfeed/tts/SpeakableItem.kt
package com.yonisirote.readmyfeed.tts

// Map content languages to concrete engine locales we actually want to request.
private val speechLanguageByContentLanguage = mapOf(
  "en" to "en-US",
  "he" to "he-IL",
)

private val hebrewCharacterPattern = Regex("[\\u0590-\\u05FF]")

data class SpeakableItem(
  val text: String,
  val authorLabel: String,
  val lang: String,
)

fun getSpeakableItemLanguage(item: SpeakableItem): String? {
  // Mixed-script posts are often mislabeled, so obvious Hebrew text wins over metadata.
  if (hebrewCharacterPattern.containsMatchIn(item.text)) {
    return "he-IL"
  }

  return speechLanguageByContentLanguage[item.lang.trim().lowercase()]
}

fun getSpeakableItemText(item: SpeakableItem): String {
  if (item.authorLabel.isBlank()) {
    return item.text.trim()
  }

  val connector = if (getSpeakableItemLanguage(item) == "he-IL") {
    "\u05d0\u05d5\u05de\u05e8:"
  } else {
    "says:"
  }

  // Use the resolved speech language here so mixed-script posts still sound natural.
  return "${item.authorLabel.trim()} $connector ${item.text.trim()}".trim()
}
